botorch_community/models/vbll_helper.py

```diff
@@ -5,531 +5,475 @@
+        return Normal(self.loc @ inp, new_scale)
 
-    def squeeze(self, idx):
+    def squeeze(self, idx: int) -> "DenseNormalPrec":
         return DenseNormalPrec(self.loc.squeeze(idx), self.tril.squeeze(idx))
 
 
-def get_parameterization(p):
+def get_parameterization(p: str):
     COV_PARAM_DICT = {
         "dense": DenseNormal,
         "dense_precision": DenseNormalPrec,
         "diagonal": Normal,
         "lowrank": LowRankNormal,
     }
-
-    try:
-        return COV_PARAM_DICT[p]
-    except KeyError:
+    if p not in COV_PARAM_DICT:
         raise ValueError(f"Invalid covariance parameterization: {p!r}")
+    return COV_PARAM_DICT[p]
 
 
-# following functions/classes are from
-# https://github.com/VectorInstitute/vbll/blob/main/vbll/layers/regression.py
-def gaussian_kl(p, q_scale):
+def gaussian_kl(p: Union[Normal, DenseNormal, LowRankNormal, DenseNormalPrec], q_scale: float) -> Tensor:
+    """
+    KL between variational posterior p (with zero-mean prior scaled by q_scale).
+    q_scale may be float or tensor; we coerce to p.mean dtype/device.
+    """
     feat_dim = p.mean.shape[-1]
-    mse_term = (p.mean**2).sum(-1).sum(-1) / q_scale
-    trace_term = (p.trace_covariance / q_scale).sum(-1)
-    logdet_term = (feat_dim * np.log(q_scale) - p.logdet_covariance).sum(-1)
-    return 0.5 * (mse_term + trace_term + logdet_term)  # currently exclude constant
+    dtype = p.mean.dtype
+    device = p.mean.device
+    q_scale_t = torch.as_tensor(float(q_scale), dtype=dtype, device=device)
+
+    mse_term = (p.mean ** 2).sum(dim=-1).sum(dim=-1) / q_scale_t
+    trace_term = (p.trace_covariance / q_scale_t).sum(dim=-1)
+    logdet_term = (feat_dim * torch.log(q_scale_t) - p.logdet_covariance).sum(dim=-1)
+    return 0.5 * (mse_term + trace_term + logdet_term)
 
 
 @dataclass
 class VBLLReturn:
-    predictive: Normal | DenseNormal
-    train_loss_fn: Callable[[torch.Tensor], torch.Tensor]
-    val_loss_fn: Callable[[torch.Tensor], torch.Tensor]
+    predictive: Union[Normal, DenseNormal, LowRankNormal, DenseNormalPrec]
+    train_loss_fn: Callable[[Tensor], Tensor]
+    val_loss_fn: Callable[[Tensor], Tensor]
 
 
 class Regression(nn.Module):
     def __init__(
         self,
-        in_features,
-        out_features,
-        regularization_weight,
-        parameterization="dense",
-        mean_initialization=None,
-        prior_scale=1.0,
-        wishart_scale=1e-2,
-        cov_rank=None,
-        clamp_noise_init=True,
-        dof=1.0,
+        in_features: int,
+        out_features: int,
+        regularization_weight: float,
+        parameterization: str = "dense",
+        mean_initialization: Optional[str] = None,
+        prior_scale: float = 1.0,
+        wishart_scale: float = 1e-2,
+        cov_rank: Optional[int] = None,
+        clamp_noise_init: bool = True,
+        dof: float = 1.0,
     ):
-        """
-        Variational Bayesian Linear Regression
-
-        Parameters
```

Comment:
It would be nice to keep these in, although BoTorch uses google-style docstrings with an "Args:" block so that things are rendered correctly by Sphinx